src/tracer.c

```diff
@@ -75,210 +75,189 @@ next_finfo(PROCESS_INFO *pi)
-	for (int j = 0; j < sizeof (long); j++)
-	{
+	for (int j = 0; j < sizeof (long); j++) {
 	    *dest++ = data.cval[j];
 	    if (data.cval[j] == 0)
 		break;
 	}
-    }
+	++i;
+    } while (*dest);
     return strdup(buf);
 }
 
-/*====================================================================================*/
-
-#if 0
-// still TODO
+#define find(arr, size, pid)						\
+({ 									\
+    size_t i = 0;							\
+    for (; i < size; ++i)						\
+    {									\
+	if (arr[i].pid == pid)						\
+	{								\
+	    break;							\
+	}								\
+    }									\
+									\
+									\
+    if(i == size)							\
+    { 									\
+        error(EXIT_FAILURE, errno, "process %d isn't in array\n", pid);	\
+    }									\
+    arr + i; 								\
+})
 
 static void
 handle_syscall(pid_t pid, const struct ptrace_syscall_info *entry,
-	       const struct ptrace_syscall_info *exit, files * buffer)
+	       const struct ptrace_syscall_info *exit)
 {
-    if (exit->exit.rval < 0)
-    {
+    if (exit->exit.rval < 0) {
 	return;			       // return on syscall failure
     }
 
-    int syscall = entry->entry.nr;
+    FILE_INFO      *finfo = next_finfo(find(pinfo, numpinfo + 1, pid));
 
-    if (syscall == SYS_open || syscall == SYS_creat)
-    {
-	puts(read_str_from_process((char *) entry->entry.args[0], pid));
-    } else if (syscall == SYS_openat)
-    {
-	puts(read_str_from_process((char *) entry->entry.args[1], pid));
+    int             syscall = entry->entry.nr;
+
+    if (syscall == SYS_open || syscall == SYS_creat) {
+	finfo->path = get_str_from_process(pid, (void *) entry->entry.args[0]);
+	finfo->purpose = entry->entry.args[1];
+    } else if (syscall == SYS_openat) {
+	finfo->path = get_str_from_process(pid, (void *) entry->entry.args[1]);
+	finfo->purpose = entry->entry.args[2];
     }
 }
 
-typedef struct state
-{
+typedef struct {
     struct ptrace_syscall_info info;
-    pid_t pid;
+    pid_t           pid;
 } state;
 
-#  define vector_name vector_state
-#  define value_type state
+#define PROCESSES_RUNNING_MAX 10240
 
-#  include "vector.h"
-
-#  undef vector_name
-#  undef value_type
-
-state *
-find(const struct vector_state *vec, pid_t pid)
-{
-    for (size_t i = 0; i < vec->size; ++i)
-    {
-	if (vec->arr[i].pid == pid)
-	{
-	    return vec->arr + i;
-	}
-    }
-
-    FATAL_ERROR_MSG("process %d isn't in children\n", pid);
-}
-
-static int
-tracer_main(pid_t pid, files * buffer)
+static void
+tracer_main(pid_t pid)
 {
-    printf("%d entered\n", pid);
     waitpid(pid, NULL, 0);
     ptrace(PTRACE_SETOPTIONS, pid, NULL,	// Options are inherited
 	   PTRACE_O_EXITKILL | PTRACE_O_TRACESYSGOOD | PTRACE_O_TRACECLONE |
 	   PTRACE_O_TRACEFORK | PTRACE_O_TRACEVFORK);
 
     struct ptrace_syscall_info info;
-    struct vector_state children;
-
-    vector_state_new(&children);
-    vector_state_reserve(&children, 16);
-
-    state temp;
+    static state    children[PROCESSES_RUNNING_MAX];
```

Comment:
No need for `children`; processes are stored in the global `pinfo`

Reply:
@zvr then i need to add an extra member of type `struct ptrace_syscall_info`